x-pack/legacy/plugins/epm/server/datasources/create.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'src/core/server/';
+import { CallESAsCurrentUser } from '../lib/cluster_access';
+import { installPipelines } from '../lib/ingest_pipelines';
+import { AssetReference } from '../../common/types';
+import { SAVED_OBJECT_TYPE_DATASOURCES } from '../../common/constants';
+import { Datasource, DatasourceAttributes } from '../../common/types';
+
+export async function createDatasource(options: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgkey: string;
+  callCluster: CallESAsCurrentUser;
+}) {
+  const { savedObjectsClient, pkgkey, callCluster } = options;
+  const toSave = await installPipelines({ pkgkey, callCluster });
```

Comment:
When a data source is created, also all its assets have to be directly installed? I'm good with this for now but I think further down there will need to be some decoupling. For example for an ML Job.

It will also be interesting on how we handle errors during the setup and how we roll back.

Reply:
Yes, for now the design is that all the work will be done when the datasource is created. I'm not entirely happy with it, but the alternative I see is to expose all the steps that are done during datasource creation as HTTP API calls and call them separately.

This would have the advantage of easier testing, and less coupled functionality, but the disadvantage that some other place (our client code?) will then need to make sure that all the steps happened and succeeded.

And yes, error handling and roll back still need some thoughts.